File: ra-idp/src/oauth2/usecases/push-authorization-request.ts
/**
 * Layer 3 — Application Logic
 *
 * /par (RFC 9126) ロジック。
 * クライアント認証済みの状態で受け取った認可パラメータを保存し、 request_uri を返す。
 */

import { randomBytes } from 'crypto'
import { OAuthError } from '../protocol/oauth-error'
import { PARRecordSchema, type PARRecord } from '../../spec-bindings/schemas'
import type { ClientRepository } from '../ports/client-repository'
import type { PARStore } from '../ports/authorization-store'

const PAR_TTL_SECONDS = 600
const REQUEST_URI_PREFIX = 'urn:ietf:params:oauth:request_uri:'

export interface PARInput {
  client_id: string
  parameters: Record<string, string>
}

export interface PARResponse {
  request_uri: string
  expires_in: number
}

export async function pushAuthorizationRequestUseCase(
  deps: {
    clientRepo: ClientRepository
    parStore: PARStore
  },
  input: PARInput,
  now: Date = new Date(),
): Promise<PARResponse> {
  const client = await deps.clientRepo.findById(input.client_id)
  if (!client) {
    throw new OAuthError('invalid_client', 'クライアント認証に失敗しました')
  }

  if (!input.parameters.code_challenge) {
    throw new OAuthError('invalid_request', 'PKCE code_challenge が必要です')
  }
  if (input.parameters.client_id && input.parameters.client_id !== input.client_id) {
    throw new OAuthError('invalid_request', 'client_id がクライアント認証と一致しません')
  }
  if (input.parameters.redirect_uri) {
    if (!client.redirect_uris.includes(input.parameters.redirect_uri)) {
      throw new OAuthError('invalid_request', 'redirect_uri が登録されていません')
    }
  }

  const requestUri = REQUEST_URI_PREFIX + randomBytes(32).toString('base64url')
  const record: PARRecord = PARRecordSchema.parse({
    request_uri: requestUri,
    client_id: input.client_id,
    parameters: input.parameters,
    issued_at: now.toISOString(),
    expires_at: new Date(now.getTime() + PAR_TTL_SECONDS * 1000).toISOString(),
    used: false,
  })
  await deps.parStore.save(record)

  return { request_uri: requestUri, expires_in: PAR_TTL_SECONDS }
}

File: ra-idp-go/internal/oauth2/usecases/push_authorization_request.go
// /par (RFC 9126 Pushed Authorization Request)
package usecases

import (
	"context"
	"time"

	"ra-idp-go/internal/oauth2/ports"
	"ra-idp-go/internal/spec"
	"ra-idp-go/internal/tenancy"
)

type PARInput struct {
	ClientID   string
	Parameters map[string]string
}

type PARResult struct {
	RequestURI string
	ExpiresIn  int
}

type PARDeps struct {
	ClientRepo          ports.OAuth2ClientRepository
	Store               ports.PARStore
	AuthzDetailTypeRepo ports.AuthorizationDetailTypeRepository
	Emit                func(spec.DomainEvent)
}

func PushAuthorizationRequest(ctx context.Context, deps PARDeps, in PARInput, now time.Time) (*PARResult, error) {
	if now.IsZero() {
		now = time.Now().UTC()
	}
	tenantID := tenancy.TenantID(ctx)
	client, err := deps.ClientRepo.FindByID(ctx, tenantID, in.ClientID)
	if err != nil {
		return nil, err
	}
	if client == nil {
		return nil, NewOAuthError("invalid_client", "未知の client_id")
	}
	// RFC 9396 — authorization_details があれば push 時点で fail-closed 検証する (ADR-050)。
	if raw := in.Parameters["authorization_details"]; raw != "" {
		details, err := ParseAuthorizationDetails(raw)
		if err != nil {
			return nil, err
		}
		if err := ValidateAuthorizationDetails(ctx, deps.AuthzDetailTypeRepo, details); err != nil {
			return nil, err
		}
	}
	id, err := generateOpaqueToken(32)
	if err != nil {
		return nil, err
	}
	requestURI := "urn:ietf:params:oauth:request_uri:" + id
	rec := &spec.PARRecord{
		TenantID:   tenantID,
		RequestURI: requestURI,
		ClientID:   in.ClientID,
		Parameters: in.Parameters,
		IssuedAt:   now,
		ExpiresAt:  now.Add(90 * time.Second), // RFC 9126 §4 推奨上限
	}
	if err := rec.Validate(); err != nil {
		return nil, err
	}
	if err := deps.Store.Save(ctx, rec); err != nil {
		return nil, err
	}
	emit(deps.Emit, &spec.PARStored{At: now, TenantID: tenantID, RequestURI: requestURI, ClientID: in.ClientID})
	return &PARResult{RequestURI: requestURI, ExpiresIn: 90}, nil
}
